picky/src/x509/pkcs7/authenticode.rs

```diff
@@ -0,0 +1,1886 @@
+                                        }
+
+                                        return Ok(());
+                                    }
+                                }
+                                UnsignedAttributeValue::CounterSign(_) => {}
+                            };
+                        }
+                    }
+                }
+
+                Err(AuthenticodeError::CertError(err))
+            }
+        }
+    }
+
+    #[cfg(feature = "ctl")]
+    fn verify_ca_certificate_against_ctl(&self, ca_name: &DirectoryName) -> AuthenticodeResult<()> {
+        use chrono::{DateTime, Duration, NaiveDate, Utc};
+        use picky_asn1::wrapper::OctetStringAsn1;
+        use std::ops::Add;
+
+        // In CTL time in a OctetString encoded as 64 bits Windows FILETIME LE
+        let time_octet_string_to_utc_time = |time: &OctetStringAsn1| -> DateTime<Utc> {
+            let since: DateTime<Utc> = DateTime::from_utc(NaiveDate::from_ymd(1601, 1, 1).and_hms(0, 0, 0), Utc);
+            since.add(Duration::seconds(
+                i64::from_le_bytes([
+                    time.0[0], time.0[1], time.0[2], time.0[3], time.0[4], time.0[5], time.0[6], time.0[7],
+                ]) / 10_000_000,
+            ))
+        };
+
+        let raw_ca_name = picky_asn1_der::to_vec(&Name::from(ca_name.clone()))?;
+        let ca_name_md5_digest = HashAlgorithm::MD5.digest(&raw_ca_name);
+
+        let ctl = CertificateTrustList::new()?;
+        let clt_entries = ctl.ctl_entries()?;
+
+        // find the CA certificate info by its md5 name digest
+        let ca_ctl_entry_attributes = clt_entries
+            .iter()
+            .find(|&ctl_entry| {
+                ctl_entry.attributes.0.iter().any(|attr| match &attr.value {
+                    CTLEntryAttributeValues::CertSubjectNameMd5HashPropId(ca_cert_md5_hash) => {
+                        match &ca_cert_md5_hash.0.first() {
+                            Some(ca_cert_md5_hash) => ca_cert_md5_hash.0 == ca_name_md5_digest,
+                            None => false,
+                        }
+                    }
+                    _ => false,
+                })
+            })
+            .ok_or(AuthenticodeError::CAIsNotTrusted)?;
+
+        // check if the CA certificate was revoked
+        if let Some(CTLEntryAttributeValues::CertDisallowedFileTimePropId(when_ca_cert_was_revoked)) =
+            ca_ctl_entry_attributes
+                .attributes
+                .0
+                .iter()
+                .find(|attr| matches!(attr.value, CTLEntryAttributeValues::CertDisallowedFileTimePropId(_)))
+                .map(|attr| &attr.value)
+        {
+            let when_ca_cert_was_revoked = when_ca_cert_was_revoked
+                .0
+                .first()
+                .expect("Asn1SetOf CertDisallowedFiletimePropId should contain exactly one value");
+
+            if when_ca_cert_was_revoked.0.is_empty() || when_ca_cert_was_revoked.0.len() < 8 {
```

Comment:
Why `len() < 8` if CertDisallowedFiletimePropId "should contain _exactly_ one value" as stated above?

Reply:
`CertDisallowedFiletimePropId` is an `Asn1SetOf<OctectStringAsn1>` which should contain exactly one value, but the one `OctectStringAsn1` value can have different length. Actually `len() < 8` is a fallback. The value always was 0 or 8 bytes long during dev-testing/researching.